os/mm/mm_heap/mm_realloc.c

```diff
@@ -381,7 +381,13 @@ FAR void *mm_realloc(FAR struct mm_heap_s *heap, FAR void *oldmem, size_t size)
 		newmem = (FAR void *)mm_malloc(heap, size);
 #endif
 		if (newmem) {
-			memcpy(newmem, oldmem, oldsize);
+			size_t copysize = oldsize - SIZEOF_MM_ALLOCNODE;
+#ifdef CONFIG_DISABLE_REALLOC_NEIGHBOR_EXTENSION
+			if (newsize <= oldsize) {
```

Comment:
what is relation with NEIGHBOR EXTENTION case, it seems should be re-calculate again for all cases

Reply:
@juitem 
`NEIGHBOR_EXTENSION enable and newsize <= oldsize` case is handled at line 158,
we need only consider `newsize <= oldsize` when NEIGHBOR_EXTENSION is disabled here.

Reply:
@juitem, zhouxinhe is right. for newsize <= oldsize case, it will be returned at line 158.

Reply:
Without this code, just use ```size```. no?

Reply:
@sunghan-chang 
Without this code, I face a heap overwritten issue.
1. heap status
ALLOC NODE 1
ALLOC NODE 2
ALLOC NODE 3
FREE NODE
2. realloc memory of ALLOC NODE 2 (NEIGHBOR_EXTENSION is disabled, or it's impossible to extend), then heap status change like this:
ALLOC NODE 1
ALLOC NODE 2 // 2nd: to be freed
ALLOC NODE 3
ALLOC NODE 4 // 1st: new allocated
FREE NODE
3. memcpy from ALLOC NODE 2 to NODE 4
if `data size` of NODE 4 is less than `total size` of NODE 2, FREE NODE area will be overwritten...

Reply:
Thank you for information. But we considered overall of realloc.
1. CONFIG_DISABLE_REALLOC_XXX_EXTENSION
 This feature is disabling extension. That means new size should be bigger than old size. But we blocked (newsize < oldsize) case as well. But there is no issue on smaller size and is not matched with config name and intention. We will not adapt it on smaller size.
2. With 1st change, no need to check smaller case at that location. Just subtract ALLOCNAME is enough and it is more simple.

Reply:
Welcome! Yes, that's a good idea! ^^

Reply:
@zhouxinhe  Anyway you found our big problem and gave a solution. Thank you very much.